### PR TITLE
 [Az.RecoveryServices.Backup] Customer Managed Key Encryption for recovery services vault

### PR DESCRIPTION
**Pull request by hiaga**
## Description

CMK for V1 vaults

## Checklist

- [Yes] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [Yes] The title of the PR is clear and informative
- [Yes] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [Yes] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [N/A] If applicable, the changes made in the PR have proper test coverage
- [Yes] For public API changes to cmdlets:
    - [Yes] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        -  https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/793
    - [Yes] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by hiaga** [comment]
@isra-fel : can you ppls merge the cmdlet review PR - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/793

File: src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md:38]: Encryption properties should be returned as part of the result of `Get-AzRecoveryServicesVaultProperty`. User should be able to get **all** the properties without needing to specify what to get.

File: src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md:88]: As we said, use consistent casing
```suggestion
### -EncryptionKeyId
```

**Review by isra-fel — CHANGES_REQUESTED** [review]

File: src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md
Comment (pvrk) [reply src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md:38]: @isra-fel : Currently, encryption properties are returned via one API and other properties are called via another API. We needed to merge them together and we don't have time for this release. We do have a plan to integrate at an API level and the same API returning all properties. Until then, my proposal is to unblock customers by having this option and then later we don't need this (either by API merging or by PS client code merging)

Hope this makes sense.

File: src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md
Comment (hiaga) [reply src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md]: fixed

File: src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md
Comment (isra-fel) [reply src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md]: I understand your intention to make it work before merging the APIs. I'm just suggesting you to take a different way to do it:
`Get-AzRecoveryServicesVaultProperty` will
- call both APIs
- conbine the key properties into the vault properties
- write the vault properties as the output

This is for 2 reasons:
1. `-EncryptionSettings` will deprecate sooner or later. Removing it will be a breaking change. Why not we don't have it at the first place.
2. Changing the output of a cmdlet dramatically by the precense of a switch parameter makes it hard to use

Comment (hiaga) [reply src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVaultProperty.md]: @isra-fel : I agree, but a small doubt - How can we modify the output type of this cmdlet (if we merge the output and create a new type)? won't it be a breaking change for customers using the existing parameter set?

File: src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVault.md
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices/help/Get-AzRecoveryServicesVault.md:55]: Could use a sample output here, for example

```suggestion
PS C:\> $vault.Identity 

IdentityType: SystemAssigned
ObjectId: xxx
```

File: src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md:24]: We talked about specifying the key by vault name + key name + (optional) key version, which should be a third parameter set. But it's OK if you choose to add that in the next sprint

File: src/RecoveryServices/RecoveryServices.Test/RecoveryServices.Test.csproj
Comment (isra-fel) [inline src/RecoveryServices/RecoveryServices.Test/RecoveryServices.Test.csproj:14]: Need to use exact same version of SDK across the solution

```suggestion
    <PackageReference Include="Microsoft.Azure.Management.RecoveryServices" Version="4.3.1-preview" />
```

File: src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md
Comment (hiaga) [reply src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md:24]: We are interested in first trying it this way, we can add it next CMK update. Also we aren't sure of hardcoding the Key url because it might use different format in service. is there a way we can construct the exact key.Id with these three values?

Comment (isra-fel) [reply src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md:24]: It's basically `$"https://{KeyVaultName}.{KeyVaultDnsSuffix}/keys/{key-name}"` with optional `"/{key-version}"` at the end.

You can get `{KeyVaultDnsSuffix}` by `DefaultContext.Environment.GetEndpoint(AzureEnvironment.Endpoint.AzureKeyVaultDnsSuffix)`

Comment (hiaga) [reply src/RecoveryServices/RecoveryServices/help/Set-AzRecoveryServicesVaultProperty.md:24]: Great we would want to add this parameter set later.

**Comment by hiaga** [comment]
@isra-fel: added the changes. pls review.

File: src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs:143]: @isra-fel  added this check to skip waiting in case Playback mode. seems like tests are still stuck. can you pls let me know the environment variable for /azp and the value it takes.

Comment (isra-fel) [reply src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs:143]: Sorry I was wrong. We didn't set a envrionment when running tests. However there's a static field [TestMockSupport.RunningMocked](https://github.com/Azure/azure-powershell-common/blob/master/src/Common/TestMockSupport.cs#L24)  you can check.

In fact, I don't know why you change all the ` TestMockSupport.Delay()` , that should work perfectly fine here

Comment (isra-fel) [reply src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs:143]: Besides, not only here, but all the "System.Threading.Sleep()` should be wrapped in an `If` clause, otherwise the tests will still time out

**Comment by isra-fel** [comment]
Also, I tried to run the tests locally, there seem to be 2 cases still failling, needing to be recorded.

TestAzureVMSetVaultProperty in RecoveryServices.Backup.Test/ScenarioTests/IaasVm/ItemTests.cs
TestRecoveryServicesVaultCRUD in RecoveryServices.Test/ScenarioTests/RecoveryServicesTests.cs

Please either record them or skip.

File: src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs
Comment (hiaga) [reply src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Jobs/WaitAzureRmRecoveryServicesBackupJob.cs:143]: @isra-fel :  included "TestMockSupport.RunningMocked" and skipped two failed tests for now. I changed  TestMockSupport.Delay() to Thread.Sleep() because while running the tests in record mode this was not putting a delay and was flooding our service with many calls. However it was working fine while running the cmdlets.

**Review by isra-fel — APPROVED** [review]

**Comment by isra-fel** [comment]
CI stuck at generation test on MacOS. Last run was sucessful. It's OK to force merge.

**Pull request merged**
